fission-server/src/extract/authority.rs

```diff
@@ -0,0 +1,143 @@
+//! Authority extractor
+
+use axum::{
+    async_trait,
+    extract::{FromRequestParts, TypedHeader},
+    headers::{authorization::Bearer, Authorization},
+    http::request::Parts,
+    RequestPartsExt,
+};
+
+use ucan::ucan::Ucan;
+
+// 🧬
+
+use crate::authority::Authority;
+use fission_core::{
+    authority,
+    authority::Error::{InvalidUcan, MissingCredentials},
+};
+
+////////////
+// TRAITS //
+////////////
+
+#[async_trait]
+impl<S> FromRequestParts<S> for Authority
+where
+    S: Send + Sync,
+{
+    type Rejection = authority::Error;
+
+    async fn from_request_parts(parts: &mut Parts, _state: &S) -> Result<Self, Self::Rejection> {
+        // Extract the token from the authorization header
+        let TypedHeader(Authorization(bearer)) = parts
+            .extract::<TypedHeader<Authorization<Bearer>>>()
+            .await
+            .map_err(|_| MissingCredentials)?;
+
+        // Decode the UCAN
+        let token = bearer.token();
+        let ucan = Ucan::try_from(token).map_err(|err| InvalidUcan {
+            reason: err.to_string(),
+        })?;
+
+        // Construct authority
+        let authority = Authority { ucan };
+
+        // Validate the authority
+        authority
+            .validate()
+            .await
+            .map(|_| authority)
+            .map_err(|reason| InvalidUcan { reason })
+    }
+}
+
+///////////
+// TESTS //
+///////////
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use axum::{
+        http::StatusCode,
+        routing::{get, Router},
+    };
+    use fission_core::authority::key_material::{generate_ed25519_material, SERVER_DID};
+    use http::Request;
+    use tower::ServiceExt;
+    use ucan::builder::UcanBuilder;
+
+    #[tokio::test]
+    async fn extract_authority() {
+        let issuer = generate_ed25519_material();
+
+        // Test if request requires a valid UCAN
+        let app: Router<(), axum::body::Body> = Router::new().route(
+            "/",
+            get(|_authority: Authority| async { axum::body::Empty::new() }),
+        );
+
+        // If a valid UCAN is given
+        let ucan = UcanBuilder::default()
+            .issued_by(&issuer)
+            .for_audience(SERVER_DID)
+            .with_lifetime(100)
+            .build()
+            .unwrap()
+            .sign()
+            .await
+            .unwrap();
+
+        let ucan_string: String = Ucan::encode(&ucan).unwrap();
+        let authed = app
+            .clone()
```

Comment:
@zeeshanlakhani I assume this is bad practice, how do I do this without `.clone()`?

Reply:
@icidasset so for tests, cloning doesn't matter :). If this was in the actual code, we'd chat otherwise.